### PR TITLE
TosaToLinalg: Support unsigned tosa.max_pool2d (FXML-4556)

### PR DESCRIPTION
**Pull request by mgehre-amd**
This PR allows to lower unsigned `tosa.max_pool2d` to linalg.
```
// CHECK-LABEL: @max_pool_ui8
func.func @max_pool_ui8(%arg0: tensor<1x6x34x62xui8>) -> tensor<1x4x32x62xui8> {
  // CHECK: builtin.unrealized_conversion_cast {{.*}} : tensor<1x6x34x62xui8> to tensor<1x6x34x62xi8>
  // CHECK: arith.constant 0
  // CHECK: linalg.pooling_nhwc_max_unsigned {{.*}} : (tensor<1x4x32x62xi8>) -> tensor<1x4x32x62xi8>
  // CHECK: builtin.unrealized_conversion_cast {{.*}} : tensor<1x4x32x62xi8> to tensor<1x4x32x62xui8>
  %0 = tosa.max_pool2d %arg0 {pad = array<i64: 0, 0, 0, 0>, kernel = array<i64: 3, 3>, stride = array<i64: 1, 1>} : (tensor<1x6x34x62xui8>) -> tensor<1x4x32x62xui8>
  return %0 : tensor<1x4x32x62xui8>
}
```
It does this by
1) Moving the type converter from mlir/lib/Conversion/TosaToLinalg/TosaToLinalgPass.cpp into `populateTosaToLinalgTypeConversion`
2) Adding materializations to the type converter using UnrealizedConversionCastOp
3) Converting the MaxPool2dConverter from OpRewriterPattern to OpConversion Pattern
4) Adjusting the padding value to the the minimum unsigned value when the max_pool is unsigned
5) Lowering to `linalg.pooling_nhwc_max_unsigned` (which uses `arith.maxui`) when the max_pool is unsigned

### CODE REVIEW

**Comment by patosgui** [comment]
LGTM

**Review by josel-amd — APPROVED** [review]
LGTM

File: mlir/lib/Conversion/TosaToLinalg/TosaToLinalgPass.cpp [hunk 1]
```diff
@@ -46,31 +46,7 @@ struct TosaToLinalg : public impl::TosaToLinalgBase<TosaToLinalg> {
...
-          SmallVector<Type> inputs;
-          if (failed(converter.convertTypes(ty.getInputs(), inputs)))
-            return std::nullopt;
-
-          SmallVector<Type> results;
-          if (failed(converter.convertTypes(ty.getResults(), results)))
-            return std::nullopt;
-
-          return FunctionType::get(ty.getContext(), inputs, results);
-        });
```
[the remark this answers is not in the document]
Comment (josel-amd) [reply mlir/lib/Conversion/TosaToLinalg/TosaToLinalgPass.cpp:73]: I had the same question here and I thought exactly what you wrote. I also saw on the test case how the function signature remains the same.

[the remark this answers is not in the document]
Comment (mgehre-amd) [reply mlir/lib/Conversion/TosaToLinalg/TosaToLinalgPass.cpp:73]: I'm considering keeping the function interfaces unchanged (i.e. signed/unsigned) while in linalg, so we can then also keep it when we go from there into emitc. We would then create an unrealized cast here, and cancel it out when we go back to emitc.

The alternative would be to convert the function signature to signless when we go into linalg, and then we would be unable to recover it when going to emitc; thus making everything signed in emitc.

[the remark this answers is not in the document]
Comment (cferry-AMD) [reply mlir/lib/Conversion/TosaToLinalg/TosaToLinalgPass.cpp:73]: I agree it helps us for our use case. As this changes the spec of the pass, is there some place this new invariant should be mentioned (or the former behavior removed from)?

**Review by cferry-AMD — APPROVED** [review]
LGTM. Let's see what upstream folks think for the interface; I'd prefer the approach proposed here but there may be a use case that needs it changed.

File: mlir/lib/Conversion/TosaToLinalg/TosaToLinalgPass.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (cferry-AMD) [reply mlir/lib/Conversion/TosaToLinalg/TosaToLinalgPass.cpp:73]: Thanks!

**Pull request merged**